statistics/shortest_time_to_get_all_singles_and_averages.rb

```diff
@@ -15,39 +19,34 @@ def query
         start_date,
         best,
         average
-      FROM Results
-      JOIN Persons person ON person.id = personId and person.subId = 1
+      FROM (
+        -- People who have single for every official event.
+        SELECT personId FROM RanksSingle GROUP BY personId HAVING COUNT(eventId) = #{Events::OFFICIAL.length}
+      ) AS all_events_people
+      JOIN (
+        -- People who have average for every official event.
+        SELECT personId FROM RanksAverage GROUP BY personId HAVING COUNT(eventId) = #{NUM_EVENTS_WITH_AVERAGES}
+      ) AS all_average_people ON all_average_people.personId = all_events_people.personId
+      JOIN Results result ON result.personId = all_events_people.personId
+      JOIN Persons person ON person.id = result.personId and person.subId = 1
       JOIN Competitions competition ON competition.id = competitionId
-      JOIN Events event ON event.id = eventId
       ORDER BY start_date
     SQL
   end
 
-  # All non-blind events and 3BLD have an average.
-  NUM_EVENTS_WITH_AVERAGES = Events::OFFICIAL.length - Events::BLD.length + 1
-
   def transform(query_results)
     query_results
       .group_by { |result| result["person_link"] }
       .map do |person_link, results|
         first_competition_date = results[0]["start_date"]
-        first_successes = results
-          .select { |result| result["best"] > 0 && Events::OFFICIAL.include?(result["event_id"]) }
-          .group_by { |result| result["event_id"] }
-          .map { |event_id, results| results.map { |result| result["start_date"] }.min }
-        first_averages = results
-          .select { |result| result["average"] > 0 && Events::OFFICIAL.include?(result["event_id"]) }
-          .group_by { |result| result["event_id"] }
-          .map { |event_id, results| results.map { |result| result["start_date"] }.min }
-        days = if first_successes.length == Events::OFFICIAL.length && first_averages.length == NUM_EVENTS_WITH_AVERAGES
-                 completion_date = (first_successes + first_averages).max
-                 (completion_date - first_competition_date).to_i
-               else
-                 nil
-               end
-        [person_link, days]
+        first_successes = %w(best average).flat_map do |type|
```

Comment:
There were two almost identical pieces of code, and turning them into something more generic wasn't too hard =)